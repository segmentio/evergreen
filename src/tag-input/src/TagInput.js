/**
 * @overview TagInput accepts multiple values that can be individually removed
 */

import React from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import cx from 'classnames'
import { Text } from '../../typography'
import { withTheme } from '../../theme'
import { majorScale } from '../../scales'
import safeInvoke from '../../lib/safe-invoke'
import Tag from './Tag'

let inputId = 1

class TagInput extends React.Component {
  static propTypes = {
    /** Whether or not the inputValue should be added to the tags when the input blurs. */
    addOnBlur: PropTypes.bool,
    /** The class name to apply to the container component. */
    className: PropTypes.string,
    /** Whether or not the input should be disabled. */
    disabled: PropTypes.bool,
    /** Props to pass to the input component. Note that `ref` and `key` are not supported. See `inputRef`. */
    inputProps: PropTypes.object,
    /**
     * Ref handler for the <input> element.
     * (input: HTMLInputElement | null) => void
     */
    inputRef: PropTypes.func,
    /**
     * Callback invoked when new tags are added.
     * Returning `false` will prevent clearing the input.
     * (values: Array) => void | false
     */
    onAdd: PropTypes.func,
    /**
     * Callback invoked when the tag values change.
     * Returning `false` will prevent clearing the input.
     * (values: Array) => void | false
     */
    onChange: PropTypes.func,
    /**
     * Callback invoked when the value of the <input> is changed. Shorthand for `inputProps={{ onChange }}`.
     * (event) => void
     */
    onInputChange: PropTypes.func,
    /**
     * Callback invoked when a tag is removed.
     * Receives value and index of removed tag.
     * (value: string | node, index: number) => void
     */
    onRemove: PropTypes.func,
    /** Value or RegExp to split on pasted text or on enter keypress */
    separator: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.instanceOf(RegExp),
      PropTypes.oneOf([false])
    ]),
    /** Provide props to tag component (actually `Badge`, for now). */
    tagProps: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    /** Controlled tag values. Each value is rendered inside a tag. */
    values: PropTypes.arrayOf(PropTypes.node)
  }

  static defaultProps = {
    addOnBlur: false,
    disabled: false,
    height: 32,
    separator: /[,\n\r]/,
    values: [],
    tagProps: {}
  }

  state = {
    inputValue: '',
    isFocused: false
  }

  id = `TagInput-${inputId++}`

  addTags = (value = '') => {
    const { onAdd, onChange, values } = this.props
    const newValues = this.getValues(value)
    let shouldClearInput = safeInvoke(onAdd, newValues)

    if (typeof onChange === 'function') {
      shouldClearInput = shouldClearInput || onChange(values.concat(newValues))
    }

    if (shouldClearInput !== false) {
      this.setState({ inputValue: '' })
    }
  }

  getValues = (inputValue = '') => {
    const { separator } = this.props

    return separator
      ? inputValue
          .split(separator)
          .map(v => v.trim())
          .filter(v => v.length > 0)
      : [inputValue]
  }

  handleBackspaceToRemove = () => {
    const { values } = this.props

    // Delete last item in values
    this.removeTagAtIndex(values.length - 1)
  }

  handleBlur = event => {
    const container = event.target

    // Use raf so that the dom has time to update `activeElement`
    requestAnimationFrame(() => {
      if (!container.contains(document.activeElement)) {
        if (this.props.addOnBlur && this.state.inputValue) {
          this.addTags(this.state.inputValue)
        }
        this.setState({ isFocused: false })
      }
    })

    safeInvoke(this.props.onBlur, event)
  }

  handleInputChange = event => {
    this.setState({ inputValue: event.target.value })
    safeInvoke(this.props.onInputChange, event)
  }

  handleInputFocus = event => {
    this.setState({ isFocused: true })
    safeInvoke(this.props.onFocus, event)
  }

  handleKeyDown = event => {
    const { selectionEnd, value } = event.target

    if (event.key === 'Enter') {
      // Prevent Enter keypresses from submitting forms since they have special powers inside TagInput
      event.preventDefault()
      this.addTags(value)
    } else if (event.key === 'Backspace' && selectionEnd === 0) {
      this.handleBackspaceToRemove(event)
    }
  }

  handleRemoveTag = event => {
    // Using data attribute to simplify callback logic -- one handler for all children
    const index = Number(
      event.currentTarget.parentElement.getAttribute('data-tag-index')
    )
    this.removeTagAtIndex(index)
  }

  maybeRenderTag = (tag, index) => {
    if (!tag) {
      return null
    }

    const { disabled, tagProps } = this.props
    const props = safeInvoke(tagProps, tag, index) || tagProps

    return (
      <Tag
        key={`${tag}:${index}`}
        data-tag-index={index}
        marginRight={majorScale(1)}
        marginY="6px"
        onRemove={disabled ? null : this.handleRemoveTag}
        isRemovable={!disabled}
        {...props}
      >
        {tag}
      </Tag>
    )
  }

  removeTagAtIndex = index => {
    const { onChange, onRemove, values } = this.props
    safeInvoke(onRemove, values[index], index)

    // Remove item at index as a new array
    const newValues = values.filter((_, i) => i !== index)
    safeInvoke(onChange, newValues)
  }

  setRef = node => {
    this.input = node
    safeInvoke(this.props.inputRef, node)
  }

  render() {
    const {
      addOnBlur,
      className,
      disabled,
      height,
      inputProps,
      inputRef,
      onAdd,
      onChange,
      onInputChange,
      onRemove,
      separator,
      tagProps,
      theme,
      values,
      ...props
    } = this.props

    const { inputValue, isFocused } = this.state

    const themedContainerClassName = theme.getTagInputClassName('default')
    const themedInputClassName = theme.getTextInputClassName('none')
    const textSize = theme.getTextSizeForControlHeight(height)
    const borderRadius = theme.getBorderRadiusForControlHeight(height)

    return (
      <Box
        aria-disabled={disabled || undefined}
        aria-activedescendant={isFocused ? this.id : undefined}
        borderRadius={borderRadius}
        className={cx(themedContainerClassName, className)}
        paddingLeft={Math.round(height / 3.2)}
        paddingRight={Math.round(height / 3.2)}
        paddingY="2px"
        {...props}
        onBlur={this.handleBlur}
      >
        {values.map(this.maybeRenderTag)}
        <Text
          is="input"
          id={this.id}
          color={disabled ? 'muted' : undefined}
          disabled={disabled}
          flexGrow="1"
          height={height - 4}
          size={textSize}
          type="text"
          value={inputValue}
          {...inputProps}
          className={themedInputClassName}
          ref={this.setRef}
          onChange={this.handleInputChange}
          onFocus={this.handleInputFocus}
          onKeyDown={this.handleKeyDown}
        />
      </Box>
    )
  }
}

export default withTheme(TagInput)
