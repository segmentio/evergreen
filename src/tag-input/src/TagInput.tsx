/**
 * @overview TagInput accepts multiple values that can be individually removed
 */

import React from 'react'
import PropTypes from 'prop-types'
import Box from 'ui-box'
import cx from 'classnames'
import { Text } from '../../typography'
import { withTheme, Theme } from '../../theme'
import { majorScale } from '../../scales'
import safeInvoke from '../../lib/safe-invoke'
import Tag from './Tag'

let inputId = 1

interface TagInputProps {
  /** Whether or not the inputValue should be added to the tags when the input blurs. */
  addOnBlur?: boolean
  /** The class name to apply to the container component. */
  className?: string
  /** Whether or not the input should be disabled. */
  disabled?: boolean
  /** The vertical size of the input */
  height?: number
  /** Props to pass to the input component. Note that `ref` and `key` are not supported. See `inputRef`. */
  inputProps?: object
  /**
   * Ref handler for the <input> element.
   * (input: HTMLInputElement | null) => void
   */
  inputRef?: (inputNode: HTMLInputElement | null) => void
  /**
   * Callback invoked when new tags are added.
   * Returning `false` will prevent clearing the input.
   */
  onAdd?: (values: any[]) => void | false
  /**
   * Callback invoked when focus on the input blurs.
   */
  onBlur?: (event: React.SyntheticEvent) => void
  /**
   * Callback invoked when the tag values change.
   * Returning `false` will prevent clearing the input.
   */
  onChange?: (values: any[]) => void | false
  /**
   * Callback invoked when the input receives focus.
   */
  onFocus?: (event: React.SyntheticEvent) => void
  /**
   * Callback invoked when the value of the <input> is changed. Shorthand for `inputProps={{ onChange }}`.
   */
  onInputChange?: (event: React.SyntheticEvent) => void
  /**
   * Callback invoked when a tag is removed.
   * Receives value and index of removed tag.
   * (value: string | node, index: number) => void
   */
  onRemove?: (value: any, index: number) => void
  /** Value or RegExp to split on pasted text or on enter keypress */
  separator?: string | RegExp | false
  /** Provide props to tag component (actually `Badge`, for now). */
  tagProps?: ((tag: any, index: number) => object) | object
  /**
   * Theme provided by ThemeProvider.
   */
  theme: Theme
  /** Controlled tag values. Each value is rendered inside a tag. */
  values: React.ReactNode[]
}

interface TagInputState {
  inputValue: string
  isFocused: boolean
}

class TagInput extends React.Component<TagInputProps, TagInputState> {
  static propTypes = {
    addOnBlur: PropTypes.bool,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    height: PropTypes.number,
    inputProps: PropTypes.object,
    inputRef: PropTypes.func,
    onAdd: PropTypes.func,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onInputChange: PropTypes.func,
    onRemove: PropTypes.func,
    separator: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.instanceOf(RegExp),
      PropTypes.oneOf([false])
    ]) as PropTypes.Requireable<string | false | RegExp>,
    tagProps: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    theme: PropTypes.object.isRequired as PropTypes.Validator<Theme>,
    values: PropTypes.arrayOf(PropTypes.node) as PropTypes.Validator<
      React.ReactNode[]
    >
  }

  static defaultProps = {
    addOnBlur: false,
    disabled: false,
    height: 32,
    separator: /[,\n\r]/,
    values: [],
    tagProps: {}
  }

  state = {
    inputValue: '',
    isFocused: false
  }

  id = `TagInput-${inputId++}`

  input: any

  addTags = (value = '') => {
    const { onAdd, onChange, values } = this.props
    const newValues = this.getValues(value)
    let shouldClearInput = safeInvoke(onAdd, newValues)

    if (typeof onChange === 'function') {
      shouldClearInput = shouldClearInput || onChange(values.concat(newValues))
    }

    if (shouldClearInput !== false) {
      this.setState({ inputValue: '' })
    }
  }

  getValues = (inputValue = '') => {
    const { separator } = this.props

    return separator
      ? inputValue
          .split(separator)
          .map(v => v.trim())
          .filter(v => v.length > 0)
      : [inputValue]
  }

  handleBackspaceToRemove = () => {
    const { values } = this.props

    // Delete last item in values
    this.removeTagAtIndex(values.length - 1)
  }

  handleBlur = event => {
    const container = event.target

    // Use raf so that the dom has time to update `activeElement`
    requestAnimationFrame(() => {
      if (!container.contains(document.activeElement)) {
        if (this.props.addOnBlur && this.state.inputValue) {
          this.addTags(this.state.inputValue)
        }

        this.setState({ isFocused: false })
      }
    })

    safeInvoke(this.props.onBlur, event)
  }

  handleInputChange = event => {
    this.setState({ inputValue: event.target.value })
    safeInvoke(this.props.onInputChange, event)
  }

  handleInputFocus = event => {
    this.setState({ isFocused: true })
    safeInvoke(this.props.onFocus, event)
  }

  handleKeyDown = event => {
    const { selectionEnd, value } = event.target

    if (event.key === 'Enter') {
      // Prevent Enter keypresses from submitting forms since they have special powers inside TagInput
      event.preventDefault()
      this.addTags(value)
    } else if (event.key === 'Backspace' && selectionEnd === 0) {
      this.handleBackspaceToRemove()
    }
  }

  handleRemoveTag = event => {
    // Using data attribute to simplify callback logic -- one handler for all children
    const index = Number(
      event.currentTarget.parentElement.getAttribute('data-tag-index')
    )
    this.removeTagAtIndex(index)
  }

  maybeRenderTag = (tag, index) => {
    if (!tag) {
      return null
    }

    const { disabled, tagProps } = this.props
    const props = safeInvoke(tagProps, tag, index) || tagProps

    return (
      <Tag
        key={`${tag}:${index}`}
        data-tag-index={index}
        marginRight={majorScale(1)}
        marginY="6px"
        onRemove={disabled ? null : this.handleRemoveTag}
        isRemovable={!disabled}
        {...props}
      >
        {tag}
      </Tag>
    )
  }

  removeTagAtIndex = index => {
    const { onChange, onRemove, values } = this.props
    safeInvoke(onRemove, values[index], index)

    // Remove item at index as a new array
    const newValues = values.filter((_, i) => i !== index)
    safeInvoke(onChange, newValues)
  }

  setRef = node => {
    this.input = node
    safeInvoke(this.props.inputRef, node)
  }

  render() {
    const {
      addOnBlur,
      className,
      disabled,
      height,
      inputProps,
      inputRef,
      onAdd,
      onChange,
      onInputChange,
      onRemove,
      separator,
      tagProps,
      theme,
      values,
      ...props
    } = this.props

    const { inputValue, isFocused } = this.state

    const themedContainerClassName = theme.getTagInputClassName('default')
    const themedInputClassName = theme.getTextInputClassName('none')
    const textSize = theme.getTextSizeForControlHeight(height!)
    const borderRadius = theme.getBorderRadiusForControlHeight(height!)

    return (
      <Box
        aria-disabled={disabled || undefined}
        aria-activedescendant={isFocused ? this.id : undefined}
        borderRadius={borderRadius}
        className={cx(themedContainerClassName, className)}
        paddingLeft={Math.round(height! / 3.2)}
        paddingRight={Math.round(height! / 3.2)}
        paddingY="2px"
        {...props}
        onBlur={this.handleBlur}
      >
        {values.map(this.maybeRenderTag)}
        <Text
          is="input"
          id={this.id}
          color={disabled ? 'muted' : undefined}
          disabled={disabled}
          flexGrow="1"
          height={height! - 4}
          size={textSize}
          type="text"
          value={inputValue}
          {...inputProps}
          className={themedInputClassName}
          ref={this.setRef}
          onChange={this.handleInputChange}
          onFocus={this.handleInputFocus}
          onKeyDown={this.handleKeyDown}
        />
      </Box>
    )
  }
}

export default withTheme(TagInput)
