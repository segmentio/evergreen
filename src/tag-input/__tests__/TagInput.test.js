import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import TagInput from '../src/TagInput'

const TEST_VALUES = ['one', 'two', 'three']
const TEST_PLACEHOLDER = 'Enter something...'

describe('<TagInput />', () => {
  it('basic snapshot', () => {
    const component = <TagInput values={TEST_VALUES} inputProps={{ placeholder: TEST_PLACEHOLDER }} />
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('onAdd', () => {
    const mockOnAdd = jest.fn()
    const newTestVal = 'Testing'
    render(<TagInput values={TEST_VALUES} onAdd={mockOnAdd} inputProps={{ placeholder: TEST_PLACEHOLDER }} />)
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), `${newTestVal}{enter}`)
    expect(screen.queryByTestId('TagInput-autocomplete-toggle')).not.toBeInTheDocument()
    expect(mockOnAdd).toHaveBeenCalledWith([newTestVal])
  })

  it('onRemove', () => {
    const mockOnRemove = jest.fn()
    render(<TagInput values={TEST_VALUES} onRemove={mockOnRemove} inputProps={{ placeholder: TEST_PLACEHOLDER }} />)
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), '{backspace}')
    const lastValueIndex = TEST_VALUES.length - 1
    expect(mockOnRemove).toHaveBeenCalledWith(TEST_VALUES[lastValueIndex], lastValueIndex)
  })

  it('onChange', () => {
    const mockOnChange = jest.fn()
    const newTestVal = 'Testing'
    render(<TagInput values={TEST_VALUES} onChange={mockOnChange} inputProps={{ placeholder: TEST_PLACEHOLDER }} />)
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), `${newTestVal}{enter}`)
    expect(mockOnChange).toHaveBeenLastCalledWith(TEST_VALUES.concat([newTestVal]))
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), '{backspace}')
    const valuesLastRemoved = TEST_VALUES.slice(0, -1)
    expect(mockOnChange).toHaveBeenLastCalledWith(valuesLastRemoved)
  })

  it('tagSubmitKey', () => {
    const mockOnAdd = jest.fn()
    const newTestVal = 'Testing'
    render(
      <TagInput
        tagSubmitKey="space"
        values={TEST_VALUES}
        onAdd={mockOnAdd}
        inputProps={{ placeholder: TEST_PLACEHOLDER }}
      />
    )
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), `${newTestVal}{space}`)
    expect(mockOnAdd).toHaveBeenCalledWith([newTestVal])
  })

  it('disabled', () => {
    render(<TagInput disabled values={TEST_VALUES} inputProps={{ placeholder: TEST_PLACEHOLDER }} />)
    TEST_VALUES.forEach(v => {
      // Checks to make sure the "X" icon is not within each tag
      expect(screen.getByText(v).children.length).toBe(0)
    })
    expect(screen.getByPlaceholderText(TEST_PLACEHOLDER)).toBeDisabled()
  })

  it('addOnBlur', () => {
    const mockOnAdd = jest.fn()
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => cb())
    const newTestVal = 'Testing'
    render(
      <TagInput
        addOnBlur
        data-testid="wrapper"
        values={TEST_VALUES}
        onAdd={mockOnAdd}
        inputProps={{ placeholder: TEST_PLACEHOLDER }}
      />
    )
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), `${newTestVal}`)
    screen.getByPlaceholderText(TEST_PLACEHOLDER).blur()
    fireEvent.blur(screen.getByPlaceholderText(TEST_PLACEHOLDER))
    expect(mockOnAdd).toHaveBeenCalledWith([newTestVal])
    window.requestAnimationFrame.mockRestore()
  })

  it('separator', () => {
    const mockOnAdd = jest.fn()
    const newTestVal = 'Testing|123'
    render(
      <TagInput separator="|" values={TEST_VALUES} onAdd={mockOnAdd} inputProps={{ placeholder: TEST_PLACEHOLDER }} />
    )
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), `${newTestVal}{enter}`)
    expect(mockOnAdd).toHaveBeenCalledWith(['Testing', '123'])
  })

  it('autocompleteItems', () => {
    const mockOnAdd = jest.fn()
    const testAutocompleteItems = ['Testing1', 'Testing2', 'Testing3', 'Other']
    const testSearch = 'Test'
    render(
      <TagInput
        autocompleteItems={testAutocompleteItems}
        values={TEST_VALUES}
        onAdd={mockOnAdd}
        inputProps={{ placeholder: TEST_PLACEHOLDER }}
      />
    )
    expect(screen.getByTestId('TagInput-autocomplete-toggle')).toBeInTheDocument()
    testAutocompleteItems.forEach(i => {
      expect(screen.queryByText(i)).not.toBeInTheDocument()
    })
    userEvent.type(screen.getByPlaceholderText(TEST_PLACEHOLDER), testSearch)
    testAutocompleteItems.forEach(i => {
      if (i.startsWith(testSearch)) {
        expect(screen.queryByText(i)).toBeInTheDocument()
      } else {
        expect(screen.queryByText(i)).not.toBeInTheDocument()
      }
    })
    userEvent.click(screen.getByText(testAutocompleteItems[0]))
    expect(mockOnAdd).toHaveBeenCalledWith([testAutocompleteItems[0]])
  })
})
