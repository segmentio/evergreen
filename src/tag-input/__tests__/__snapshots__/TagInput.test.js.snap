// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagInput /> basic snapshot 1`] = `
<div
  className="css-1defwtr ub-algn-itms_center ub-dspl_inline-flex ub-flx-wrap_wrap ub-pst_relative ub-pb_2px ub-pt_2px ub-bg-clr_white ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-b-btm_1px-solid-d8dae5 ub-b-lft_1px-solid-d8dae5 ub-b-rgt_1px-solid-d8dae5 ub-b-top_1px-solid-d8dae5 ub-box-szg_border-box"
  onBlur={[Function]}
>
  <strong
    className="css-1rdjc1y ub-fnt-sze_11-5px ub-f-wght_400 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-dspl_inline-flex ub-box-szg_border-box ub-ver-algn_middle ub-h_16px ub-pb_6px ub-pt_6px ub-pl_8px ub-pr_4px ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-txt-algn_center ub-txt-deco_none ub-txt-trns_none ub-bg-clr_edeff5 ub-algn-itms_center ub-mt_6px ub-mb_6px ub-ml_8px ub-mr_8px"
    data-tag-index={0}
  >
    one
    <svg
      className="css-c4d79v ub-ml_4px ub-w_12px ub-h_12px ub-box-szg_border-box"
      data-icon="cross"
      onClick={[Function]}
      viewBox="0 0 16 16"
    >
      <path
        d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
        fillRule="evenodd"
      />
    </svg>
  </strong>
  <strong
    className="css-1rdjc1y ub-fnt-sze_11-5px ub-f-wght_400 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-dspl_inline-flex ub-box-szg_border-box ub-ver-algn_middle ub-h_16px ub-pb_6px ub-pt_6px ub-pl_8px ub-pr_4px ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-txt-algn_center ub-txt-deco_none ub-txt-trns_none ub-bg-clr_edeff5 ub-algn-itms_center ub-mt_6px ub-mb_6px ub-ml_8px ub-mr_8px"
    data-tag-index={1}
  >
    two
    <svg
      className="css-c4d79v ub-ml_4px ub-w_12px ub-h_12px ub-box-szg_border-box"
      data-icon="cross"
      onClick={[Function]}
      viewBox="0 0 16 16"
    >
      <path
        d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
        fillRule="evenodd"
      />
    </svg>
  </strong>
  <strong
    className="css-1rdjc1y ub-fnt-sze_11-5px ub-f-wght_400 ub-ln-ht_16px ub-ltr-spc_0 ub-fnt-fam_b77syt ub-color_474d66 ub-dspl_inline-flex ub-box-szg_border-box ub-ver-algn_middle ub-h_16px ub-pb_6px ub-pt_6px ub-pl_8px ub-pr_4px ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-txt-algn_center ub-txt-deco_none ub-txt-trns_none ub-bg-clr_edeff5 ub-algn-itms_center ub-mt_6px ub-mb_6px ub-ml_8px ub-mr_8px"
    data-tag-index={2}
  >
    three
    <svg
      className="css-c4d79v ub-ml_4px ub-w_12px ub-h_12px ub-box-szg_border-box"
      data-icon="cross"
      onClick={[Function]}
      viewBox="0 0 16 16"
    >
      <path
        d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
        fillRule="evenodd"
      />
    </svg>
  </strong>
  <div
    className="ub-flx-grow_1 ub-dspl_inline-block ub-box-szg_border-box"
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="TagInput-1-label"
      aria-owns={null}
      role="combobox"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls={null}
        aria-invalid={false}
        aria-labelledby="TagInput-1-label"
        autoComplete="off"
        className="css-2radon ub-w_auto ub-fnt-fam_b77syt ub-b-btm_1px-solid-transparent ub-b-lft_1px-solid-transparent ub-b-rgt_1px-solid-transparent ub-b-top_1px-solid-transparent ub-otln_iu2jf4 ub-txt-deco_none ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-ln-ht_16px ub-fnt-sze_12px ub-color_474d66 ub-pl_12px ub-pr_12px ub-tstn_n1akt6 ub-h_28px ub-bg-clr_transparent ub-f-wght_400 ub-ltr-spc_0 ub-box-szg_border-box"
        disabled={false}
        id="TagInput-1"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Enter something..."
        spellCheck={true}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
